File: bindings/csharp/Kevy.Client/Transaction.cs
// Transactions: MULTI / EXEC / DISCARD + WATCH (contract §3.12). Remote-
// only — the embedded backend rejects multi/watch/unwatch with Unsupported
// (single-connection embedded access is already serial). Wire flow: MULTI →
// +OK; each queued command → +QUEUED; EXEC → array of N typed replies. A
// WATCH violation makes EXEC return Nil (aborted). A Transaction abandoned
// without exec/discard sends an implicit DISCARD via Dispose/DisposeAsync so
// the socket isn't left in MULTI mode.

namespace Kevy;

public sealed partial class KevyClient
{
    /// <summary>WATCH keys for optimistic concurrency (≥1). Remote-only.</summary>
    public void Watch(params KevyBytes[] keys)
    {
        if (keys.Length == 0) throw new KevyInvalidInputException("WATCH needs at least one key");
        SimpleOk(RequireRemote("WATCH").Request(Argv.Cmd("WATCH", keys.Raw())));
    }

    /// <summary>UNWATCH every WATCH on this connection. Remote-only.</summary>
    public void Unwatch() => SimpleOk(RequireRemote("UNWATCH").Request([Argv.S("UNWATCH")]));

    /// <summary>Start a MULTI block. Remote-only.</summary>
    public Transaction Multi()
    {
        var conn = RequireRemote("MULTI/EXEC");
        SimpleOk(conn.Request([Argv.S("MULTI")]));
        return new Transaction(conn);
    }

    internal static void SimpleOk(Reply r)
    {
        if (r.Kind == ReplyKind.Simple && r.Str() == "OK") return;
        throw r.Kind == ReplyKind.Error ? Err.ReplyError(r) : Err.Unexpected(r);
    }
}

/// <summary>One in-flight MULTI block over a remote connection (§3.12).
/// Dispose (or Discard) to send the implicit DISCARD; dropping a live
/// transaction without it is a bug.</summary>
public sealed class Transaction : IDisposable, IAsyncDisposable
{
    private readonly RespConnection _conn;
    private bool _live = true;

    internal Transaction(RespConnection conn) => _conn = conn;

    /// <summary>Queue one raw command (verb + args); expects +QUEUED.</summary>
    public Transaction Queue(params KevyBytes[] parts)
    {
        if (parts.Length == 0) throw new KevyInvalidInputException("Transaction.Queue needs at least a verb");
        QueueArgv(parts.Raw());
        return this;
    }

    /// <summary>Queue SET key value.</summary>
    public Transaction Set(KevyBytes key, KevyBytes value) => QueueChain([Argv.S("SET"), key.Raw, value.Raw]);
    /// <summary>Queue GET key.</summary>
    public Transaction Get(KevyBytes key) => QueueChain([Argv.S("GET"), key.Raw]);
    /// <summary>Queue DEL key…</summary>
    public Transaction Del(params KevyBytes[] keys) => Require(keys, "Del").Pipe(Argv.Cmd("DEL", keys.Raw()));
    /// <summary>Queue EXISTS key…</summary>
    public Transaction Exists(params KevyBytes[] keys) => Require(keys, "Exists").Pipe(Argv.Cmd("EXISTS", keys.Raw()));
    /// <summary>Queue INCR key.</summary>
    public Transaction Incr(KevyBytes key) => QueueChain([Argv.S("INCR"), key.Raw]);
    /// <summary>Queue INCRBY key delta.</summary>
    public Transaction IncrBy(KevyBytes key, long delta) => QueueChain([Argv.S("INCRBY"), key.Raw, Argv.I(delta)]);
    /// <summary>Queue MGET key…</summary>
    public Transaction MGet(params KevyBytes[] keys) => Require(keys, "MGet").Pipe(Argv.Cmd("MGET", keys.Raw()));
    /// <summary>Queue MSET key value…</summary>
    public Transaction MSet(params KevyBytes[] pairs)
    {
        if (pairs.Length == 0 || pairs.Length % 2 != 0)
            throw new KevyInvalidInputException("Transaction.MSet needs a non-empty even key/value list");
        return QueueChain(Argv.Cmd("MSET", pairs.Raw()));
    }

    /// <summary>EXEC; per-command replies. A WATCH abort (Nil) collapses to
    /// an empty list (legacy).</summary>
    public IReadOnlyList<Reply> Exec()
    {
        var r = FinishExec();
        return r.Kind switch
        {
            ReplyKind.Array => r.Items,
            ReplyKind.Nil or ReplyKind.Null => Array.Empty<Reply>(),
            ReplyKind.Error => throw Err.ReplyError(r),
            _ => throw Err.Unexpected(r),
        };
    }

    /// <summary>EXEC; null when a WATCH violation aborted it.</summary>
    public IReadOnlyList<Reply>? ExecWatched()
    {
        var r = FinishExec();
        return r.Kind switch
        {
            ReplyKind.Array => r.Items,
            ReplyKind.Nil or ReplyKind.Null => null,
            ReplyKind.Error => throw Err.ReplyError(r),
            _ => throw Err.Unexpected(r),
        };
    }

    /// <summary>EXEC → a typed reply cursor; a WATCH abort is a Protocol error.</summary>
    public TransactionReplies ExecTyped()
    {
        var r = FinishExec();
        return r.Kind switch
        {
            ReplyKind.Array => new TransactionReplies(r.Items),
            ReplyKind.Nil or ReplyKind.Null => throw new KevyProtocolException("transaction aborted by WATCH"),
            ReplyKind.Error => throw Err.ReplyError(r),
            _ => throw Err.Unexpected(r),
        };
    }

    /// <summary>EXEC → a typed cursor, or null on WATCH abort.</summary>
    public TransactionReplies? ExecWatchedTyped()
    {
        var r = FinishExec();
        return r.Kind switch
        {
            ReplyKind.Array => new TransactionReplies(r.Items),
            ReplyKind.Nil or ReplyKind.Null => null,
            ReplyKind.Error => throw Err.ReplyError(r),
            _ => throw Err.Unexpected(r),
        };
    }

    /// <summary>Abandon the queued commands.</summary>
    public void Discard()
    {
        if (!_live) return;
        _live = false;
        KevyClient.SimpleOk(_conn.Request([Argv.S("DISCARD")]));
    }

    private Reply FinishExec()
    {
        _live = false;
        return _conn.Request([Argv.S("EXEC")]);
    }

    private Transaction QueueChain(byte[][] argv) { QueueArgv(argv); return this; }
    private Transaction Pipe(byte[][] argv) { QueueArgv(argv); return this; }
    private Transaction Require(KevyBytes[] keys, string op) =>
        keys.Length == 0 ? throw new KevyInvalidInputException($"Transaction.{op} needs at least one key") : this;

    private void QueueArgv(byte[][] argv)
    {
        var r = _conn.Request(argv);
        if (r.Kind == ReplyKind.Simple && r.Str() == "QUEUED") return;
        throw r.Kind == ReplyKind.Error ? Err.ReplyError(r) : Err.Unexpected(r);
    }

    /// <summary>Implicit DISCARD if still live (§3.12 drop semantics).</summary>
    public void Dispose()
    {
        if (!_live) return;
        _live = false;
        try { _conn.Request([Argv.S("DISCARD")]); } catch (KevyException) { }
    }

    /// <inheritdoc cref="Dispose"/>
    public ValueTask DisposeAsync() { Dispose(); return ValueTask.CompletedTask; }
}

/// <summary>A typed cursor over a successful EXEC's per-command replies
/// (§4.10). Each Next* consumes one reply; a variant mismatch is a Protocol
/// error (the cursor still advances so ExpectEmpty stays meaningful).</summary>
public sealed class TransactionReplies
{
    private readonly IReadOnlyList<Reply> _items;
    private int _pos;

    internal TransactionReplies(IReadOnlyList<Reply> items) => _items = items;

    /// <summary>How many replies are un-consumed.</summary>
    public int Remaining => _items.Count - _pos;

    /// <summary>Error if any replies remain (an arity gate).</summary>
    public void ExpectEmpty()
    {
        if (Remaining != 0) throw new KevyProtocolException($"transaction reply cursor has {Remaining} un-consumed replies");
    }

    /// <summary>Pop the next reply as-is.</summary>
    public Reply Raw()
    {
        if (_pos >= _items.Count) throw new KevyProtocolException("exhausted");
        return _items[_pos++];
    }

    /// <summary>Expect +OK.</summary>
    public void NextOk()
    {
        var v = Raw();
        if (!(v.Kind == ReplyKind.Simple && v.Str() == "OK")) throw Mismatch("Simple(OK)", v);
    }

    /// <summary>+OK → true, Nil → false (for SET … NX/XX).</summary>
    public bool NextOkOrNil()
    {
        var v = Raw();
        if (v.Kind == ReplyKind.Simple && v.Str() == "OK") return true;
        if (v.IsNil) return false;
        throw Mismatch("Simple(OK) or Nil", v);
    }

    /// <summary>Expect an integer.</summary>
    public long NextInt()
    {
        var v = Raw();
        return v.Kind == ReplyKind.Int ? v.Integer : throw Mismatch("Int", v);
    }

    /// <summary>Expect a bulk (or Nil → null).</summary>
    public byte[]? NextBulk()
    {
        var v = Raw();
        if (v.Kind == ReplyKind.Bulk) return v.Bytes;
        if (v.IsNil) return null;
        throw Mismatch("Bulk or Nil", v);
    }

    /// <summary>Expect an array of Bulk/Nil (MGET), in order.</summary>
    public IReadOnlyList<byte[]?> NextArrayOfBulks()
    {
        var v = Raw();
        if (v.IsNil) return Array.Empty<byte[]?>();
        if (v.Kind != ReplyKind.Array) throw Mismatch("Array", v);
        var outp = new List<byte[]?>(v.Items.Count);
        foreach (var it in v.Items)
        {
            if (it.Kind == ReplyKind.Bulk) outp.Add(it.Bytes);
            else if (it.IsNil) outp.Add(null);
            else throw Mismatch("Array element Bulk/Nil", it);
        }
        return outp;
    }

    /// <summary>Expect any simple string (e.g. +PONG).</summary>
    public byte[] NextSimple()
    {
        var v = Raw();
        return v.Kind == ReplyKind.Simple ? v.Bytes : throw Mismatch("Simple", v);
    }

    private static KevyProtocolException Mismatch(string want, Reply got) =>
        new($"transaction reply mismatch: expected {want}, got {got.Shape()}");
}

File: bindings/python/kevy/_transaction.py
"""Transactions: MULTI / EXEC / DISCARD (+ WATCH), contract §3.12.

Remote-only. A ``Transaction`` abandoned without exec/discard MUST send an
implicit DISCARD so the socket is not left in MULTI mode — Python has no
deterministic destructors, so this is a context manager (``with``) and a
best-effort ``__del__``; dropping a live transaction without close is a bug.
``TransactionReplies`` is a typed cursor over a successful EXEC (§4.10).
"""

from __future__ import annotations

from typing import List, Optional

from ._decode import Bytesish, to_bytes
from ._errors import InvalidInputError, ProtocolError, error_from_reply_text
from ._reply import Reply, ReplyKind


class Transaction:
    """One in-flight MULTI block over a remote connection."""

    def __init__(self, conn):
        self._conn = conn
        self._live = True

    def __enter__(self) -> "Transaction":
        return self

    def __exit__(self, *exc) -> None:
        self.close()

    def __del__(self):
        try:
            self.close()
        except Exception:
            pass

    # --- queueing -------------------------------------------------------

    def queue(self, *parts: Bytesish) -> "Transaction":
        """Queue one raw command (verb + args); expects +QUEUED."""
        if not parts:
            raise InvalidInputError("queue needs at least a verb")
        return self._queue([to_bytes(p) for p in parts])

    def set(self, key: Bytesish, value: Bytesish) -> "Transaction":
        return self._queue([b"SET", to_bytes(key), to_bytes(value)])

    def get(self, key: Bytesish) -> "Transaction":
        return self._queue([b"GET", to_bytes(key)])

    def delete(self, *keys: Bytesish) -> "Transaction":
        if not keys:
            raise InvalidInputError("delete needs at least one key")
        return self._queue([b"DEL", *[to_bytes(k) for k in keys]])

    def exists(self, *keys: Bytesish) -> "Transaction":
        if not keys:
            raise InvalidInputError("exists needs at least one key")
        return self._queue([b"EXISTS", *[to_bytes(k) for k in keys]])

    def incr(self, key: Bytesish) -> "Transaction":
        return self._queue([b"INCR", to_bytes(key)])

    def incr_by(self, key: Bytesish, delta: int) -> "Transaction":
        return self._queue([b"INCRBY", to_bytes(key), str(int(delta)).encode()])

    def mget(self, *keys: Bytesish) -> "Transaction":
        if not keys:
            raise InvalidInputError("mget needs at least one key")
        return self._queue([b"MGET", *[to_bytes(k) for k in keys]])

    def mset(self, *pairs: Bytesish) -> "Transaction":
        pb = [to_bytes(p) for p in pairs]
        if not pb or len(pb) % 2 != 0:
            raise InvalidInputError("mset needs a non-empty even key/value list")
        return self._queue([b"MSET", *pb])

    def _queue(self, argv: List[bytes]) -> "Transaction":
        r = self._conn.request(argv)
        if r.kind is ReplyKind.SIMPLE and r.data == b"QUEUED":
            return self
        if r.is_error():
            raise error_from_reply_text(r.data)
        raise ProtocolError(f"expected +QUEUED, got {r.shape()}")

    # --- execution ------------------------------------------------------

    def _finish(self) -> Reply:
        self._live = False
        return self._conn.request([b"EXEC"])

    def exec(self) -> List[Reply]:
        """EXEC → per-command replies. A WATCH abort (Nil) collapses to []
        (legacy); prefer exec_watched for new code."""
        r = self._finish()
        if r.kind is ReplyKind.ARRAY:
            return r.items
        if r.is_nil():
            return []
        if r.is_error():
            raise error_from_reply_text(r.data)
        raise ProtocolError(f"EXEC: unexpected {r.shape()}")

    def exec_watched(self) -> Optional[List[Reply]]:
        """None on WATCH abort; [] on empty success."""
        r = self._finish()
        if r.kind is ReplyKind.ARRAY:
            return r.items
        if r.is_nil():
            return None
        if r.is_error():
            raise error_from_reply_text(r.data)
        raise ProtocolError(f"EXEC: unexpected {r.shape()}")

    def exec_typed(self) -> "TransactionReplies":
        """A typed reply cursor; a WATCH abort is a ProtocolError."""
        r = self._finish()
        if r.kind is ReplyKind.ARRAY:
            return TransactionReplies(r.items)
        if r.is_nil():
            raise ProtocolError("transaction aborted by WATCH")
        if r.is_error():
            raise error_from_reply_text(r.data)
        raise ProtocolError(f"EXEC: unexpected {r.shape()}")

    def exec_watched_typed(self) -> Optional["TransactionReplies"]:
        r = self._finish()
        if r.kind is ReplyKind.ARRAY:
            return TransactionReplies(r.items)
        if r.is_nil():
            return None
        if r.is_error():
            raise error_from_reply_text(r.data)
        raise ProtocolError(f"EXEC: unexpected {r.shape()}")

    def discard(self) -> None:
        """Abandon the queued commands."""
        if not self._live:
            return
        self._live = False
        r = self._conn.request([b"DISCARD"])
        if r.is_error():
            raise error_from_reply_text(r.data)

    def close(self) -> None:
        """Send an implicit DISCARD if still live, so the socket is not left
        in MULTI mode. Idempotent."""
        if self._live:
            self._live = False
            try:
                self._conn.request([b"DISCARD"])
            except Exception:
                pass


class TransactionReplies:
    """A typed cursor over a successful EXEC's per-command replies (§4.10).
    Each ``next_*`` consumes one reply; a variant mismatch is a
    ProtocolError and the cursor still advances so ``expect_empty`` stays
    meaningful."""

    def __init__(self, items: List[Reply]):
        self._items = items
        self._pos = 0

    @property
    def remaining(self) -> int:
        return len(self._items) - self._pos

    def expect_empty(self) -> None:
        left = self.remaining
        if left != 0:
            raise ProtocolError(f"transaction reply cursor has {left} un-consumed replies")

    def raw(self) -> Reply:
        if self._pos >= len(self._items):
            raise ProtocolError("transaction reply cursor exhausted")
        v = self._items[self._pos]
        self._pos += 1
        return v

    def next_ok(self) -> None:
        v = self.raw()
        if v.kind is ReplyKind.SIMPLE and v.data == b"OK":
            return
        raise _mismatch("Simple(OK)", v)

    def next_ok_or_nil(self) -> bool:
        v = self.raw()
        if v.kind is ReplyKind.SIMPLE and v.data == b"OK":
            return True
        if v.is_nil():
            return False
        raise _mismatch("Simple(OK) or Nil", v)

    def next_int(self) -> int:
        v = self.raw()
        if v.kind is ReplyKind.INT:
            return v.integer
        raise _mismatch("Int", v)

    def next_bulk(self) -> Optional[bytes]:
        v = self.raw()
        if v.kind is ReplyKind.BULK:
            return v.data
        if v.is_nil():
            return None
        raise _mismatch("Bulk or Nil", v)

    def next_array_of_bulks(self) -> List[Optional[bytes]]:
        v = self.raw()
        if v.kind is ReplyKind.ARRAY:
            out: List[Optional[bytes]] = []
            for it in v.items:
                if it.kind is ReplyKind.BULK:
                    out.append(it.data)
                elif it.is_nil():
                    out.append(None)
                else:
                    raise _mismatch("Array element Bulk/Nil", it)
            return out
        if v.is_nil():
            return []
        raise _mismatch("Array", v)

    def next_simple(self) -> bytes:
        v = self.raw()
        if v.kind is ReplyKind.SIMPLE:
            return v.data
        raise _mismatch("Simple", v)


def _mismatch(want: str, got: Reply) -> ProtocolError:
    return ProtocolError(f"transaction reply mismatch: expected {want}, got {got.shape()}")
